protocol/morse/context.go

```diff
@@ -15,58 +15,107 @@ import (
-	return availableEndpoints, nil
+	return endpoints, nil
 }
 
-// HandleServiceRequest satisfies the gateway package's ProtocolRequestContext interface.
+// HandleServiceRequest handles incoming service request.
 // It uses the supplied payload to send a relay request to an endpoint, and verifies and returns the response.
 func (rc *requestContext) HandleServiceRequest(payload protocol.Payload) (protocol.Response, error) {
+	// TODO_IMPROVE(@adshmh): use the same pattern for hydrated loggers in any packages with large number of logger fields.
+	hydratedLogger := rc.getHydratedLogger("HandleServiceRequest")
+
 	if rc.selectedEndpoint == nil {
+		// Internal error: no endpoint selected, record an observation but no sanctions
+		// as no endpoint was contacted
+		hydratedLogger.Error().Msg("no endpoint has been selected.")
+
+		rc.recordEndpointObservation(
+			endpoint{}, // No endpoint selected
+			protocolobservations.MorseEndpointErrorType_MORSE_ENDPOINT_ERROR_INTERNAL,
+			"no endpoint has been selected, endpoint selection has not been performed or has failed",
+			protocolobservations.MorseSanctionType_MORSE_SANCTION_UNSPECIFIED, // No sanction as no endpoint was contacted
+		)
+
 		return protocol.Response{}, fmt.Errorf("HandleServiceRequest: no endpoint has been selected on service %s", rc.serviceID)
 	}
 
 	morseEndpoint, err := getEndpoint(rc.selectedEndpoint.session, rc.selectedEndpoint.Addr())
 	if err != nil {
-		return protocol.Response{},
-			fmt.Errorf("HandleServiceRequest: error matching the selected endpoint %s against session's nodes: %w", rc.selectedEndpoint.Addr(), err)
+		// Internal error: endpoint not in session, record an observation but no sanctions
+		// as this is an internal error, not an endpoint issue
+		hydratedLogger.Error().Err(err).Msg("endpoint not found in session.")
+
+		rc.recordEndpointObservation(
+			*rc.selectedEndpoint,
+			protocolobservations.MorseEndpointErrorType_MORSE_ENDPOINT_ERROR_INTERNAL,
+			fmt.Sprintf("error matching endpoint against session's nodes: %v", err),
+			protocolobservations.MorseSanctionType_MORSE_SANCTION_UNSPECIFIED, // No sanction as this is an internal error
+		)
+
+		return protocol.Response{}, NewEndpointNotInSessionError(string(rc.selectedEndpoint.Addr()))
 	}
 
 	output, err := rc.sendRelay(
-		string(rc.serviceID),
+		hydratedLogger,
 		morseEndpoint,
 		rc.selectedEndpoint.session,
 		rc.selectedEndpoint.app.aat,
-		// TODO_IMPROVE: chain-specific timeouts
+		// TODO_FUTURE(@adshmh): support service-specific timeouts, passed from the Qos instance.
 		0, // SDK to use the default timeout.
 		payload,
 	)
 
+	// Record any errors that occurred during relay
+	if err != nil {
+		endpointErrorType, recommendedSanctionType := classifyRelayError(rc.logger, err)
+
+		hydratedLogger.Error().
+			Str("error_type", endpointErrorType.String()).
+			Str("sanction_type", recommendedSanctionType.String()).
+			Err(err).
+			Msg("relay error occurred.")
+
+		rc.recordEndpointObservation(
```

Comment:
For the future, do we not want to record positive observations? 

i.e. when err == nil?

If not - ignore this comment.
If you see any value for it - TODO_FUTURE?